dockstore-client/src/main/java/io/dockstore/client/cli/nested/BaseLanguageClient.java

```diff
@@ -0,0 +1,305 @@
+        this.entry = entryVal;
+        this.isLocalEntry = localEntry;
+        this.yamlParameterFile = yamlFile;
+        this.jsonParameterFile = jsonFile;
+        this.wdlOutputTarget = outputTarget;
+        this.uuid = notificationUUID;
+    }
+
+    /**
+     * Common code to setup and launch a pipeline
+     * @return Exit code of process
+     */
+    public long launchPipeline(String entryVal, boolean localEntry, String yamlFile, String jsonFile, String outputTarget, String notificationUUID) throws ApiException {
+        // Initialize client with some launch information
+        setLaunchInformation(entryVal, localEntry, yamlFile, jsonFile, outputTarget, notificationUUID);
+
+        // Load up Docker images
+        abstractEntryClient.loadDockerImages();
+
+        // Select the appropriate parameter file
+        selectedParameterFile = selectParameterFile();
+
+        // Setup the launcher (Download dependencies)
+        launcher.initialize();
+
+        // Setup notifications
+        setupNotifications();
+
+        // Setup temp directory and download files
+        downloadFiles();
+
+        try {
+            // Provision the input files
+            provisionedParameterFile = provisionInputFiles();
+        } catch (ApiException ex) {
+            if (abstractEntryClient.getEntryType().toLowerCase().equals("tool")) {
+                exceptionMessage(ex, "The tool entry does not exist. Did you mean to launch a local tool or a workflow?",
+                        ENTRY_NOT_FOUND);
+            } else {
+                exceptionMessage(ex, "The workflow entry does not exist. Did you mean to launch a local workflow or a tool?",
+                        ENTRY_NOT_FOUND);
+            }
+        } catch (Exception ex) {
+            exceptionMessage(ex, ex.getMessage(), GENERIC_ERROR);
+        }
+
+        // Update the launcher with references to the files to be launched
+        launcher.setFiles(localPrimaryDescriptorFile, importsZipFile, provisionedParameterFile, selectedParameterFile, workingDirectory);
+
+        try {
+            // Attempt to run launcher
+            executeEntry();
+
+            // Provision the output files if run is successful
+            provisionOutputFiles();
+        } catch (ApiException ex) {
+            exceptionMessage(ex, ex.getMessage(), API_ERROR);
+        }  catch (IOException ex) {
+            exceptionMessage(ex, ex.getMessage(), IO_ERROR);
+        } catch (Exception ex) {
+            exceptionMessage(ex, ex.getMessage(), GENERIC_ERROR);
+        }
+
+        notificationsClient.sendMessage(NotificationsClient.COMPLETED, true);
+
+        return 0;
+    }
+
+    /**
+     * Creates a working directory and downloads descriptor files
+     * @param type CWL or WDL
+     * @return Pair of downloaded primary descriptor and zip file
+     */
+    public Triple<File, File, File> initializeWorkingDirectoryWithFiles(ToolDescriptor.TypeEnum type) {
```

Comment:
This is a style/opinion thing, but I would rather see a class/interface here instead of a Triple; otherwise it's hard to know which File is which.